lib/web/recordingplayback.go

```diff
@@ -0,0 +1,677 @@
+	// eventTypeStop indicates the stop of a response of events
+	eventTypeStop responseType = 2
+	// eventTypeError indicates an error
+	eventTypeError responseType = 3
+	// eventTypeSessionStart indicates session started
+	eventTypeSessionStart responseType = 4
+	// eventTypeSessionPrint contains terminal output
+	eventTypeSessionPrint responseType = 5
+	// eventTypeSessionEnd indicates session ended
+	eventTypeSessionEnd responseType = 6
+	// eventTypeResize indicates terminal resize
+	eventTypeResize responseType = 7
+	// eventTypeScreen contains terminal screen state
+	eventTypeScreen responseType = 8
+	// eventTypeBatch indicates a batch of events
+	eventTypeBatch responseType = 9
+)
+
+const (
+	// maxRequestRange is the maximum allowed time range for a request
+	maxRequestRange = 10 * time.Minute
+	// requestHeaderSize is the size of the request header (event type, start time, end time, request ID, and current screen flag)
+	requestHeaderSize = 22
+	// responseHeaderSize is the size of the response header (event type, timestamp, data size, and request ID)
+	responseHeaderSize = 17
+)
+
+// recordingPlayback manages session event streaming
+type recordingPlayback struct {
+	ctx              context.Context
+	cancel           context.CancelFunc
+	clt              events.SessionStreamer
+	sessionID        string
+	logger           *slog.Logger
+	mu               sync.Mutex
+	cancelActiveTask context.CancelFunc
+	ws               *websocket.Conn
+	writeChan        chan []byte
+
+	stream struct {
+		sync.Mutex
+		eventsChan    <-chan apievents.AuditEvent
+		errorsChan    <-chan error
+		lastEndTime   int64
+		bufferedEvent apievents.AuditEvent
+	}
+
+	terminal struct {
+		sync.Mutex
+		vt          vt10x.Terminal
+		currentTime int64
+	}
+}
+
+// fetchRequest represents a request for session events.
+type fetchRequest struct {
+	requestType          requestType
+	startOffset          int64
+	endOffset            int64
+	requestID            int
+	requestCurrentScreen bool
+}
+
+// sessionEvent represents a single session event with its type, timestamp, and data.
+type sessionEvent struct {
+	eventType responseType
+	timestamp int64
+	data      []byte
+}
+
+func (h *Handler) recordingPlaybackWs(
+	w http.ResponseWriter,
+	r *http.Request,
+	p httprouter.Params,
+	sctx *SessionContext,
+	cluster reversetunnelclient.Cluster,
+	ws *websocket.Conn,
+) (interface{}, error) {
+	sessionID := p.ByName("session_id")
+	if sessionID == "" {
+		return nil, trace.BadParameter("missing session ID in request URL")
+	}
+
+	defer ws.Close()
+
+	ctx := r.Context()
+	clt, err := sctx.GetUserClient(ctx, cluster)
+	if err != nil {
+		data := []byte(err.Error())
+
+		totalSize := responseHeaderSize + len(data)
+		buf := make([]byte, totalSize)
+
+		encodeEvent(buf, 0, eventTypeError, 0, data, 0)
+
+		if err := ws.WriteMessage(websocket.BinaryMessage, buf); err != nil {
+			h.logger.ErrorContext(ctx, "failed to send event", "session_id", sessionID, "error", err)
+		}
+
```

Comment:
should we properly handle the websocket close dance?

ws.Close won't trigger the send close frame, followed by wait for the close frame response. Without them, very fast websockets could lose frames or frames not delivered to client